### PR TITLE
Added Append(Stringbuilder,int,int) and Append(StringBuilder) overlaod apis

### PR DESCRIPTION
**Pull request by Anipik**
Fixes  https://github.com/dotnet/corefx/issues/4928
Tests Pr- https://github.com/dotnet/corefx/pull/26231
Previous Pr-  dotnet/coreclr#8615

Most of the things are same as the previous PR.

### CODE REVIEW

**Comment by Anipik** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

**Comment by Anipik** [comment]
@danmosemsft can you take a look at this ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+                    AppendCore(value, 0, value.Length);
+                else
+                    Append(value.ToString());
+            }
+            return this;
+        }
+
+        public StringBuilder Append(StringBuilder value, int startIndex, int count)
+        {
+            if (startIndex < 0)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Please use ThrowHelper instead of newing up an exception inline. See elsewhere. Probably the existing code in this class should use ThrowHelper, at least in the "hot" methods (maybe you could do that in another commit)

See reasoning for this:
https://github.com/dotnet/corefx/issues/26253#issuecomment-356736809

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 2]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: it is not important but I find it slightly easier to read if the equals true case is first. @AndyAyersMS does it matter for code gen if we put the common branch second? or does it pick the ordering itself?
```c#
               if (value == this)
                    Append(value.ToString());
               else
                    AppendCore(value, 0, value.Length);
```

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 3]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+            }
+            return this;
+        }
+
+        private void AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            int newLength = Length + count;
+            if ((uint)newLength > (uint)m_MaxCapacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_SmallMaxCapacity);
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: this is the wrong string, take a look in the resx.

should be
throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_Capacity);

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 4]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+                if (startIndex == 0 && count == 0)
+                {
+                    return this;
+                }
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (count > value.Length - startIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_Index);
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: this says index is out of range and refers to count. it could be either. maybe use ArgumentOutOfRange_IndexCount and use nameof(value) ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 2 above)
Comment (AndyAyersMS) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: It should not matter. Pick whichever version is most readable.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 5]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+                unsafe
+                {
+                    fixed (char* destinationPtr = &m_ChunkChars[m_ChunkLength])
+                    {
+                        while (curDestIndex > 0)
+                        {
+                            int chunkEndIndex = sourceEndIndex - chunk.m_ChunkOffset;
+                            if (chunkEndIndex >= 0)
+                            {
+                                if (chunkEndIndex > chunk.m_ChunkLength)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: `chunkEndIndex = Math.Min(chunkEndIndex, chunk.m_ChunkLength);`

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 6]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+                if (length == 0)
+                {
+                    ExpandByABlock(count);
+                    length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
+                }
+
+                StringBuilder chunk = value;
+                int sourceEndIndex = startIndex + length;
+                int curDestIndex = length;
+                unsafe
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: This unsafe block is identical to the same one in ToString(). I would extract a method they can share
`private static unsafe char* CopyToBuffer(StringBuilder chunk, char* destinationPtr, int startIndex, int length)`

Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: I think the body of that should be able to itself reuse the exissting
`private static unsafe void ThreadSafeCopy(char* sourcePtr, char[] destination, int destinationIndex, int count)`

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 7]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+
+                                    // Check that we will not overrun our boundaries. 
+                                    if ((uint)(chunkCount + curDestIndex) <= (uint)length && (uint)(chunkCount + chunkStartIndex) <= (uint)sourceArray.Length)
+                                    {
+                                        fixed (char* sourcePtr = &sourceArray[chunkStartIndex])
+                                            string.wstrcpy(destinationPtr + curDestIndex, sourcePtr, chunkCount);
+                                    }
+                                    else
+                                    {
+                                        throw new ArgumentOutOfRangeException(nameof(chunkCount));
```
Comment (Anipik) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: @danmosemsft  not able to find any appropriate helper function for this. Should I choose ThrowArgumentOutOfRangeException, ThrowArgumentOutOfRange_IndexException() ?

Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Either is fine .. this will never be hit except in broken apps, I think.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 8]
```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+                    AppendCore(value, startIndex, count);
+                else
+                    Append(value.ToString(startIndex, count));
+            }
+            return this;
+        }
+
+        private void AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            int newLength = Length + count;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Debug.Assert(value != this) ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 6 above)
Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Should I change this for  ToString() also ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 1 above)
Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Should I change this for existing append overloads ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 6 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: @jkotas is reuse here better than avoiding a function call?

Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: All these methods and loops with pointers passed around are unnecessary security risk. 

I would make this method 100% safe and just call `Append(char[] value, int startIndex, int charCount)` in a loop here.

If you want to go extra mile and save some of the unnecessary argument validation overhead, change the core append method to `Append(ReadOnlySpan<char>)` and make other methods including this one call it.

Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: BTW: `Append(ReadOnlySpan<char>)` would be a good public API to add - I expect it will be no-brainer to approve.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 9]
```diff
@@ -790,6 +790,109 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+            {
+                ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
+            }
+
+            if (count > 0)
+            {
+                if (value == this)
+                    Append(value.ToString(startIndex, count));             
+                else
+                    AppendCore(value, startIndex, count);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Nit: It should be slightly better to do `return AppendCore` here and other similar places in this file because of it lets the JIT to employ [tailcall](https://en.wikipedia.org/wiki/Tail_call) optimization and not worry about keeping `this` pointer alive.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 6 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: > @jkotas is reuse here better than avoiding a function call?

I see what you meant now. My suggestion with Append would not work/is unrelated.

I would optimize for minimizing/eliminating unsafe code. The StringBuilder code is not easy to reason about. We had number of security issues in it over the years. Reducing code duplication would be secondary for me.

Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Maybe something like this?

```
```
int newLength = Length + count;
if ((uint)newLength > (uint)m_MaxCapacity)
{
    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
}

while (count > 0)
{
    int length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
    if (length == 0)
    {
        ExpandByABlock(count);
        length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
    }

    value.CopyTo(startIndex, new Span<char>(m_ChunkChars, m_ChunkLength, length), count);

    startIndex += length;
    count -= length;
}
```

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 10]
```diff
@@ -790,6 +790,109 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                if (value == this)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Handing of the `value == this` case can be in AppendCore.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: We do the manual optimization using ThrowHelper in hot small methods or generic methods because of it is where it matters the most. We do not do it in "regular" methods like this because of it is extra pain to maintain.

Hopefully, we will eventually get optimization step in the build process that does the transformation of throw to ThrowHelper everywhere and we will not need to ever worry about it again.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 6 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: CopyTo has same loop as what  you have here. Calling CopyTo from here avoids the code duplication - makes the code smaller I think.

There is probably a cleanup that can be done around ThreadSafeCopy - delete the one that take pointer. I would recommend doing that as separate PR if you are interested in it.

Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: I can do the cleanup regarding the ThreadSafeCopy and toString(int,int) in another PR

Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: @jkotas which one of these is preferable for toString(int,int). We are using extra memory in first one but there is no unsafe code in this.
```
AssertInvariants();
char[] destination = new char[length];
this.CopyTo(startIndex, new Span<char>(destination), length);
return new string(destination);
```
```
AssertInvariants();
 string result = string.FastAllocateString(length);
  unsafe
   {
         fixed (char* destinationPtr = result)
    {
           this.CopyTo(startIndex, new Span<char>(destinationPtr, length), length);
            return result;
       }
}
```

Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: You certainly do not want to allocate temporary `char[]`.

You second option would be ok.

If you wanted to avoid unsafe code altogether, you would write it as:
```
return string.Create(length,
   (Source: this, StartIndex: startIndex),
   (span, state) => state.Source.CopyTo(state.StartIndex, span, span.Length));
```

The C# looks better, but it is a bit slower and quite a bit bigger code. I think I like your second option with a tiny bit of unsafe code better in this particular case.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 11]
```diff
@@ -828,18 +784,14 @@ public StringBuilder Append(StringBuilder value, int startIndex, int count)
...
-                    AppendCore(value, startIndex, count);
-            }
-            return this;
+            return AppendCore(value, startIndex, count);
         }
 
-        private void AppendCore(StringBuilder value, int startIndex, int count)
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
         {
+            if (count > 0 && value == this)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: Nit: The `value == this` condition should be first because of it is the one more likely to be false.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: You may want to undo the use of throw helper here to be consistent with the overall style. These methods are neither generic or hot.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 11 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Why is that important? Per Andy it doesn't matter for codegen?
https://github.com/dotnet/coreclr/pull/15786#discussion_r161060827

**Comment by Anipik** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 11 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: If you have a condition like `if (rarely_false && rarely_true)`, it is always better for perf to write it as `if (rarely_true && rarely_false)` - assuming that both conditions are cheap and side-effect free.

It does not make a huge difference ... just a nit.

**Comment by Anipik** [comment]
@danmosemsft @jkotas can you take a final look at this ?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 11 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Ah of course, I misunderstood I thought you meant swapping the if and else clauses.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 12]
```diff
@@ -790,6 +749,75 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+            return AppendCore(value, startIndex, count);
+        }
+
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            if (value == this && count > 0)
+                return Append(value.ToString(startIndex, count));
+
+            int newLength = Length + count;
+            Debug.Assert(value != this);
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: seems you shouldn't have this assert anymore as you allow value of this. The assert blocks appending yourself with zero count, which should work (no-op)

should be a test for that..

Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]:
```
if (count == 0)
{
     return this;
}
```
I added this check similar to string overloads

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 13]
```diff
@@ -790,6 +749,75 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                return AppendCore(value, 0, value.Length);
+            }
+            return this;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: if I do sb.Append((StringBuilder)null) shouldn't it throw?

also should `return AppendCore` be outside the if, so it can be the last line and maybe get tailcall optimization (as the common case)

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 12 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: would this work
```c#
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
 +        {
 +            if (value == this)
 +                return Append(value.ToString(startIndex, count));
```
ie, simply stop checking for 0.

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 14]
```diff
@@ -790,6 +749,79 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+            }
+
+            return AppendCore(value, startIndex, count);
+        }
+
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            if (value == this && count > 0)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: my comment got lost? you should be able to remove the `count > 0` check since `sb.ToString(startIndex, 0)` is legal

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 13 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: > if I do sb.Append((StringBuilder)null) shouldn't it throw?

I was wondering about that as well ...

File: src/mscorlib/shared/System/Text/StringBuilder.cs [hunk 15]
```diff
@@ -790,6 +749,79 @@ public StringBuilder Append(string value, int startIndex, int count)
...
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_GenericPositive);
+            }
+
+            if (value == null)
+            {
+                if (startIndex == 0 && count == 0)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Text/StringBuilder.cs]: similar, if value is null, shouldn't it throw?
or if not, why does startIndex matter if count is 0?
seems simpler to just throw for null unless there's some scenario for it

Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: I kept this for consistency. We are doing similar behavior for other overloads

Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Which ones?

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 13 above)
Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: It will return this. similar behavior to append(string)

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 15 above)
Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: https://github.com/Anipik/coreclr/blob/ee5ecc8be2f16746e975640df381f6fe7a4e580d/src/mscorlib/shared/System/Text/StringBuilder.cs#L725
https://github.com/Anipik/coreclr/blob/ee5ecc8be2f16746e975640df381f6fe7a4e580d/src/mscorlib/shared/System/Text/StringBuilder.cs#L619

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 13 above)
Comment (Anipik) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: https://github.com/Anipik/coreclr/blob/ee5ecc8be2f16746e975640df381f6fe7a4e580d/src/mscorlib/shared/System/Text/StringBuilder.cs#L690

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 15 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: Fair enough.

File: src/mscorlib/shared/System/Text/StringBuilder.cs (hunk 13 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/StringBuilder.cs]: OK, consistency is good. Please make sure there's tests for all these..
